Inject "additional-request-params" into the SecureGrantMap produced by  ClientCredentialsTokenRequestContext::getGrant

frehov opened the pull request:
- Added `getAdditionalRequestParams` into the `ClientCredentialsConfiguration` interface
- Added `additionalRequestParams` to the `ClientCredentialsConfigurationProperties`, including getter and setter.
- Added `additionalRequestParams` to `ClientCredentialsTokenRequestContext` 
- In `ClientCredentialsTokenRequestContext`, inject extra parameters when the `ClientCredentialsGrant` is converted to a map
- Added a test to verify that the binding of the added configuration properties are valid

Todo(?)
- [ ]  Update documentation regarding `additional-request-params` on client credentials grant oauth flow

closes #751

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/ClientCredentialsConfiguration.java:
```diff
@@ -48,4 +49,7 @@
     Optional<ClientCredentialsHeaderTokenPropagatorConfiguration> getHeaderPropagation();
+
+    @NonNull
+    Map<String, String> getAdditionalRequestParams();
```
This probably needs to be a `Map<String, Object>` to allow more flexibility.

sdelamo commented on security-oauth2/src/test/groovy/io/micronaut/security/oauth2/client/clientcredentials/ClientCredentialsConfigurationSpec.groovy:
```diff
@@ -54,4 +54,27 @@ class ClientCredentialsConfigurationSpec extends Specification {
+        noExceptionThrown()
+
+        expect:
+        configuration.getClientCredentials().get().getAdditionalRequestParams().get("audience") == "test"
```
You can rewrite this as: 
        
```groovy
configuration.getClientCredentials().get().getAdditionalRequestParams().get("audience") == "test"
```

as: 

```groovy
"test" == configuration.clientCredentials.get().additionalRequestParams.get("audience") 
```

frehov replied:
done

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/configuration/OauthClientConfigurationProperties.java:
```diff
@@ -277,6 +280,8 @@ public void setRevocation(RevocationEndpointConfigurationProperties revocation)
 
         private HeaderTokenPropagatorConfigurationProperties headerPropagation;
 
+        private Map<String, String> additonalRequestParams = Collections.emptyMap();
```
This probably needs to be a `Map<String, Object> `to allow more flexibility.

frehov replied:
Just curious as to the need for flexibility here?, as it feeds directly into the `SecureGrantMap` which is typed as `Map<String, String>`

sdelamo replied:
you are correct. Let's keep a `Map<String, String>` as you suggest.

sdelamo commented on security-oauth2/src/main/java/io/micronaut/security/oauth2/configuration/OauthClientConfigurationProperties.java:
```diff
@@ -435,6 +440,16 @@ public String getHeaderName() {
+            return additonalRequestParams;
+        }
+
+        public void setAdditionalRequestParams(@MapFormat(transformation = MapFormat.MapTransformation.FLAT) Map<String, String> additionalRequestParams) {
```
It misses javadoc. 

Configuration Reference is populated with javadoc of ConfigurationProperites setters. Thus, it is important to provide good Javadoc here. 

Run `./gradlew docs` and you will see the configuration reference documentation being generated.

frehov replied:
Javadoc added, please let me know if there's anything else that should be added.

Should there be an example for the Auth0 use-case maybe?

sdelamo replied:
> Should there be an example for the Auth0 use-case maybe?
 

We will change Auth0 Client Credentials guide to use this feature once micronaut security 3.1 is included in the Micronaut bom core

 https://github.com/micronaut-projects/micronaut-guides/pull/513

sdelamo reviewed (changes requested):
Thanks for the contribution: 

-  this is a new feature it should use as `base` the `master` branch. It should be included in the next minor version of `micronaut-security` module

See different comments in the reviews.

Looking forward for this change. It will make life easier for users.

sdelamo commented:
>  Update documentation regarding additional-request-params on client credentials grant oauth flow

This table generates automatically with the javadoc, see comment in the PR review. 

<img width="868" alt="Screenshot 2021-09-13 at 09 46 17" src="https://user-images.githubusercontent.com/864788/133044308-27bdbd93-e7f5-4042-b444-184ffe414084.png">

frehov commented:
@sdelamo could you advise on how I get rid of the commit from the PR that is not related to my changes?
I rebased on master and changed the base branch for which to merge into.

sdelamo commented:
@frehov can you fix the checkstyle ?

```
./gradlew :security-oauth2:checkstyleMain

> Task :security-oauth2:checkstyleMain FAILED
[ant:checkstyle] [ERROR] /Users/sdelamo/github/micronaut-projects/micronaut-security/security-oauth2/src/main/java/io/micronaut/security/oauth2/configuration/OauthClientConfigurationProperties.java:448:9: Fields and methods should be before inner classes. [InnerTypeLast]
[ant:checkstyle] [ERROR] /Users/sdelamo/github/micronaut-projects/micronaut-security/security-oauth2/src/main/java/io/micronaut/security/oauth2/configuration/OauthClientConfigurationProperties.java:454: First sentence should end with a period. [JavadocStyle]
[ant:checkstyle] [ERROR] /Users/sdelamo/github/micronaut-projects/micronaut-security/security-oauth2/src/main/java/io/micronaut/security/oauth2/configuration/OauthClientConfigurationProperties.java:458:9: Fields and methods should be before inner classes. [InnerTypeLast]

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':security-oauth2:checkstyleMain'.
> Checkstyle rule violations were found. See the report at: file:///Users/sdelamo/github/micronaut-projects/micronaut-security/security-oauth2/build/reports/checkstyle/main.html
  Checkstyle files with violations: 1
  Checkstyle violations by severity: [error:3]```

frehov commented:
@sdelamo I've rectified the checkstyle issue, but I'm seeing some weird behaviour from spotless and checkstyle.
I'm on windows, just so that's said, and spotless want's to change all files til `CRLF`, while checkstyle is configured to only accept `LF`.
my git config for `core.autocrlf` is `input`

Is this a known issue on windows?

sdelamo commented:
> I'm on windows, just so that's said, and spotless want's to change all files til `CRLF`, while checkstyle is configured to only accept `LF`.
> my git config for `core.autocrlf` is `input`
> 
> Is this a known issue on windows?
> 
> edit: I've tried setting the `core.eol` setting in my gitconfig to `lf` for the micronaut repos I have locally, but no dice.
> edit2: seems like spotless is looking for a `.gitattributes` file and then falling back to `PLATFORM_NATIVE`, It should've picked up the `core.eol` but doesn't

Not sure if anyone in the core development team uses Windows. We use either MacOs or Linux. Thus ,this maybe an issue we are not aware of.

frehov commented:
My workaround was adding a `.gitattributes` at the top level with the following, content.
```
* text eol=lf
*.bat text eol=crlf
```
with that, spotless asserted `lf` endings, and checkstyle was happy as well
